Efficient object store for the AiiDA repository (#11)
aiidateam/AEP: 003_efficient_object_store_for_repository/readme.md

- When performing "bulk" operations on many files (two common cases are when exporting a portion of the DB,
  or when performing a backup), accessing hundreds of thousands of files is extremely slow. Disk caches
  (in Linux, for instance) make this process much faster, but this works only as long as the system has a lot
  of RAM, and the files have been accessed recently. Note that this is not an issue of AiiDA. Even just running
  `rsync` on a folder with hundreds of thousands of files (not recently accessed) can take minutes or even hours
  just to check if the files need to be updated, while if the same content is in a single big file, the operation
  would take much less (possibly even less than a second for a lot of small files). As a note, this is the reason
  why at the moment we provide a backup script to perform incremental backups of the repository (checking the
  timestamp in the AiiDA database and only transferring node repository folders of new nodes), but the goal would be to rely instead on standard tools like `rsync`.

## Proposed Enhancement 
The goal of this project is to have a very efficient implementation of an "object store" that:

> - works directly on a disk folder;

Reviewer: With "works directly on a disk folder", do you mean that it is compatible with a folder file system as the backbone? And it should as the bare minimum support this?

Author: I tried to clarify. I mean it directly writes to a folder on a "standard" disk, without the need of going through a running server (as mentioned in the point right below).

- ideally, does not require a service to be running in the background,

>   to avoid to have to ask users to run yet another service to use AiiDA;

Reviewer: Do we exclude the users that have a proper object store running that is integrated into their workflow? Or do we offer such integrations in a different AEP?

Reviewer: Improving the current performance is nice. But how do we foresee a proper object store integration that gives other benefits? Also, it should be easy to select the repositories we use. E.g. this should in principle be abstracted from AiiDA. We should make sure that if we include a custom light object store like this, that it does not block future direct implementation of a proper object store.

Reviewer: One solution would be to follow the principles of e.g. `rclone`. E.g. there is an interface between AiiDA and whatever repositories the interface supports. The `disk-objectstore` should possibly be put into this interface.

Author: Integration with a proper object store is left out of this AEP, because as I mention (see also discussion below) the actual problem we're trying to solve is latency in retrieving many small files, and actual object stores are much worse than the filesystem on this regard.

Investigation on how to integrate with an "actual" object store can be followed [in this issue](https://github.com/giovannipizzi/disk-objectstore/issues/17).
But unfortunately I don't see how this can be more efficient than just using the disk...

Reviewer: Ok. Let us think carefully if this can block or make a more proper direct object store integration more complicated in the future. For this it would be nice to have some feedback from someone that works rather deeply with Swift, maybe some people from SwiftStack or similar.

Author: I see your point. I think at the moment we want to have at least a good and well performing repository for every user. We can consider as an *optional* extension something requiring a server, for those that require maximal performance. I think that anyway the current API is a step forward, because it introduces a very clear API, that can definitely be applied also to other storage backends (e.g. it provides primitives to store and retrieve data given a key, and also to perform bulk operations).

Reviewer: Sounds great.

- and addresses a number of performance issues mentioned above and discussed more in detail below.

**NOTE**:  This AEP does not address the actual implementation of the object store in AiiDA, but
rather the implementation of an object store to solve a number of performance issues.
The description of the integration with AiiDA will be discussed in a different AEP
(see [PR #7](https://github.com/aiidateam/AEP/pull/7) for some preliminary discussion, written before this AEP).

## Detailed Explanation 
The goal of this AEP is to define some design decisions for a library (an "object store", to be used internally by AiiDA) that,
given a stream of bytes, stores it as efficiently as possible somewhere, assigns a unique key (a "name") to it,
and then allows to retrieve it with that key.

The libary also supports efficient bulk write and read operations
to cover the cases of bulk operations in AiiDA (exporting, importing,
backups).

Here we describe some design criteria for such an object store.
We also provide an implementation that follows these criteria in the

> [disk-objectstore](https://github.com/giovannipizzi/disk-objectstore)

Reviewer: Can this be integrated with e.g. `rclone`, ref. comments in https://github.com/aiidateam/aiida-core/issues/335?

Author: Mentioned rclone now

repository, where efficiency has also been benchmarked.

### Performance guidelines and requirements

> #### Packing and loose objects

Reviewer: We should be very careful here. We should try to lean on existing solutions that is used by the object store community.

Author: Indeed, maybe the name "object store" is not the correct one - I tried to clarify that what I actually mean is what is referred to as a "key-value" store

  Current experience (with AiiDA) shows that it is actually not so good to use two
  levels of nesting, that was employed to avoid to have too many files in the same folder. And anyway the core idea of this implementationis that when there are too many loose objects, the user will pack them.
  The number of characters in the first part can be chosen in the library, but a good compromise 
  after testing is 2 (the default, and also the value used internally by git).

- Loose objects are actually written first to a sandbox folder (in the same filesystem),
  and then moved in place with the correct UUID only when the file is closed, with an atomic operation.
  This should prevent having leftover objects if the process dies.
  In this way, we can rely simply on the fact that an object present in the folder of loose objects signifies that the object exists, without needing a centralised place to keep track of them.
  
  Having the latter would be a potential performance bottleneck, as if there are many concurrent
  writers, the object store must guarantee that the central place is kept updated.

> - Packing can be triggered by the user periodically, whenever the user wants.

Reviewer: What does packing in this context mean? Bundle objects or compression?

Author: Clarified

  Instead, it is *not* required that packing can be performed in parallel by multiple processes
  (on the contrary, the current implementation actually tries to prevent multiple processes trying to perform
  write operations on packs concurrently).

- In practice, this operation takes all loose objects and puts them in a controllable number
  of packs. The name of the packs is given by the first few letters of the UUID
  (by default: 2, so 256 packs in total; configurable). A value of 2 is a good balance
  between the size of each pack (on average, ~4GB/pack for a 1TB repository) and
  the number of packs (having many packs means that, even when performing bulk access,
  many different files need to be open, which slows down performance).

- Pack files are just concatenation of bytes of the packed objects. Any new object
  is appended to the pack (thanks to the efficiency of opening a file for appending).
  The information for the offset and length of each pack is kept in a single SQLite
  database for the whole set of objects, as we describe below.

> - Packed objects can optionally be compressed. Note that compression is on a per-object level.

Author: No, compression is done fully "transparently" for the user. See new text comments in c9f86f7

  So, even when dealing with huge files, these never fill the RAM (e.g. when reading
  or writing a multi-GB file, the memory usage has been tested to be capped at ~150MB).
  Convenience methods are available, anyway, to get directly an object content, if
  the user wants this for simplicty, and knows that the content fits in RAM.

#### Further design choices

- Each given object will get a random UUID as a key (its generation cost is negligible, about
  4 microseconds per UUID).
  It's up to the caller to track this into a filename or a folder structure.
  The UUID is generated by the implementation and cannot be passed from the outside.
  This guarantees random distribution of objects in packs, and avoids to have to
  check for objects already existing (that can be very expensive).

- Pack naming and strategy is not determined by the user. Anyway it would be difficult
  to provide easy options to the user to customize the behavior, while implementing
  a packing strategy that is efficient. Moreover, with the current packing strategy,
  it is immediate to know in which pack to check without having to keep also an index

>   of the packs (this, however, would be possible to implement in case we want to extend the behavior,

Author: Indeed, it should be very easy, and the users wouldn't even notice. Sentence added in c9f86f7

  since anyway we have an index file). But at the moment it does not seem necessary.

- For each object, the SQLite database contains the following fields, that can be considered
  to be the "metadata" of the packed object: its key (`uuid`), the `offset` (starting
  position of the bytestream in the pack file), the `length` (number of bytes to read),

>   a boolean `compressed` flag, meaning if the bytestream has been zlib-compressed,

Author: As I now mention in c9f86f7 you can always perform compression yourself (as the user of the library) and store the compressed version. But the advantage of having the OS do it for you is that you don't have to worry about compression. You just store the data you are interested into, and depending on the settings of the repository, or more correctly on the flags with which you will call the packing, packed objects will end up (possibly) being stored as compressed, but you can keep using the object store as usual, and you will always get back the uncompressed version, independent of how the object is actually stored, which compression library was used, ...

On the second point: currently, the compression is zlib, but there are strong suggestions to move to more modern libraries (xz, or snappy, see [e.g. the discussion in this issue](https://github.com/giovannipizzi/disk-objectstore/issues/14))

  return the results, and close again the file). Also for bulk requests, anyway, this
  would prevent making very few DB requests (unless you keep all files open, that
  again is not feasible).

- We decided to use SQLite because of the following reasons:
  - it's a database, so it's efficient in searching for the metadata
    of a given object;
  - it does not require a running server;
  - in [WAL mode](https://www.sqlite.org/wal.html), allows many concurrent readers and one writer,
    useful to allow to continue normal operations by many Unix processes during repacking;
  - we access it via the SQLAlchemy library that anyway is already
    a dependency of AiiDA, and is actually the only dependency of the
    current object-store implementation.
  - it's quite widespread and so the libary to read and write should be reliable.

- Deletion can just occur efficiently as either a deletion of the loose object, or
  a removal from the index file (if the object is already packed). Later repacking of the packs can be used to recover
  the disk space still occupied in the pack files.

- The object store does not need to provide functionality to modify

>   a node. In AiiDA files of nodes are typically only added, and once

Reviewer 2: ```suggestion
  an object. In AiiDA, files of nodes are typically only added, and once
```

Author: Fixed, thanks

- Appending content to a single file does not prevent the Linux disk cache to work efficiently.
  Indeed, the caches are per blocks/pages in linux, not per file.
  Concatenating to files does not impact performance on cache efficiency. What is costly is opening a file as the filesystem
  has to provide some guarantees e.g. on concurrent access.
  As a note, seeking a file to a given position is what one typically does when watching a 
  video and jumping to a different section.

### Why a custom implementation of the library
We have been investigating if existing codes could be used for the current purpose.

However, in our preliminary analysis we couldn't find an existing robust software that was satisfying all criteria.
In particular:

- existing object storage implementations (e.g. Open Stack Swift, or others that typically provide a
  S3 or similar API) are not suitable since 1) they require a server running, and we would like to
  avoid the complexity of asking users to run yet another server, and most importantly 2) they usually work 
  via a REST API that is extremely inefficient when retrieving a lot of small objects (latencies can

>   be even of tens of ms, that is clearly unacceptable if we want to retrieve millions of objects).

Reviewer: This depends wildly on the system setup and how we connect to it. Using cloud solutions is slow, but on-prem object store solutions are indeed rather fast. Of course, they are not designed for performance (but there are solutions, like MinIO) but offer other benefits which would be nice. E.g. access control on object level, replication, integration into larger existing backends at institutions, companies etc. (metadata searches etc.), no file system limit, provides unique id automatically, scalable, high-availability, flexibility and so on. Also, at some point running a large AiiDA infrastructure on regular file systems is going to be a limiting factor, especially in multi user scenarios.

Author: I think MinIO advertises performance, but this just in the throughput rate (with impressing TB/s). But we're not really interested in that, rather on latency to get a single object. They don't provide any number for that (or at least I couldn't find), but I'd bet is not way better than any other object-store solution, and anyway orders of magnitude slower than getting a file from disk

  implementation ([dulwich](https://github.com/dulwich/dulwich)). We have been benchmarking it, but
  the feeling is that it is designed to do something else, and adapting to our needs might not be worth it.
  Some examples of issues we envision: managing re-packing (dulwich can only pack loose objects, but not repack existing packs); this can be done via git itself, but then we need to ensuring that when repacking, objects are not garbage-collected and deleted because they are not referenced within git;
  it's not possible (apparently) to decide if we want to compress an object or not (e.g., in case we want maximum performance at the cost of disk space), but they are always compressed.
  Also we did not test concurrency access and packing of the git repository, which requires some stress test to assess if it works.

- One possible solution could be a format like HDF5. However we need a very efficient hash-table like access
  (given a key, get the object), while HDF5 is probably designed to append data to the last dimension of 
  a multidimensional array. Variables exist, but (untested) probably they cannot scale well
  at the scale of tens of millions.

> - One option we didn't investigate is the mongodb object store. Note that this would require running a server

Reviewer: We should leave a comment about using e.g. `rclone` as a common interface.

Author: Mentioned rclone now

  though (but could be acceptable if for other reasons it becomes advantageous).

> Finally, as a note, we stress that an efficient implementation in about 1'000 lines of code has been

Reviewer: Fully fledged object store solutions have a very different angle of approach and utility/functionality set and is a different kind of beast. We should be a bit careful how we use the term object store as many users expect much more from that than just being able to bundle files/objects.

Author: Indeed - I think this is my fault, I should have probably called this a key-value store

  give higher guarantees, be more robust, and make it simpler to guarantee that data is not corrupted even for concurrent access.

## Pros and Cons 

### Pros
* Is very efficient also with hundreds of thousands of objects (bulk reads can read all objects in a couple of seconds).
* Is rsync-friendly, so one can suggest to use directly rsync for backups instead of custom scripts.
* An implementation exists, seems to be very efficient, and is
  relatively short (and already has tests with 100% coverage,
  including concurrent writes, reads, and one repacking process, and checking on three platforms: linux, mac os, and windows).
* It does not add any additional python depenency to AiiDA, and it
  does not require a service running.
* It can be used also for the internal format of AiiDA export files,
  and this should be very efficient, even to later retrieve just
  a subset of the files.
* Implements compression in a transparent way.
* It shouldn't be slower than the current AiiDA implementation in writing
  during normal operation, since objects are stored as loose objects.
  Actually, it might be faster for nodes without files, as no disk
  access will be needed anymore.

### Cons

> * Extreme care is needed to convince ourselves that there are no

Reviewer: Corner cases typically do not appear until this is battle tested by a larger community in different areas of application. Hence, solutions applied to a larger community are always going to be considered more stable, even though we have 100% test coverage. That only means we test the code that is written, not that the code actually handles all cases.

Author: I agree - I added a comment on this, and I fully agree that 100% coverage is not enough! I am already aware of a few things that need to be fixed in the code already.

  bugs and no risk of corrupting or losing the users' data.
* Object metadata must be tracked by the caller in some other database.
  Similarly, deduplication via SHA hashes is not implemented and need to be
  implemented by the caller.

> * It is not possible anymore to access directly the folder of a node by opening

Reviewer: Yes, maybe this should be a pro instead.

Author: Probably yes!

  As a note, seeking a file to a given position is what one typically does when watching a 
  video and jumping to a different section.

> - Packing in general, at this stage, is left to the user. We can decide (at the object-store level, or probably

Reviewer: What does packing here mean? Compression or bundling of several objects/entities? This is a bit difficult to know and I suggest we specify this from get go in the text.

Author: Hi @espenfl , I explained [at the beginning of the section](https://github.com/aiidateam/AEP/pull/11/commits/774c8d426f0a9236e0e78ba1f8f9f4fd81bb690d#diff-34bba23dcb01fde03df9d5e8bfbbcaeeR81-R83):

> The core idea behind the object store implementation is that, instead of writing a lot of small files, these are packed into a few big files, and an index of some sort keeps track of where each object is in the big file. 

Let me know if this is not clear yet.

I will get back to your comments soon.
